### ISSUE TITLE
Add client context to model objects

### ISSUE THREAD

**Issue by dylanmcreynolds** [SciCatProject/pyscicat#13]
In the early days of pyscicat, I debated two different ways to structure the API and model classes. I chose to use very  thin models based on pydantic that exist mostly to carry data and provide validation. The `ScicatClient` object contains all of the communication functionality to talk to the Scicat server. This was easy and cheap and fulfilled the original strong need of scicat: ingestion.

Talking with @nitrosx, ESS would like to use pyscicat in more of a data access scenario. Here, they would like to do something like:

```python

for datablock in dataset.datblocks():
  for file in datablock.files():
     print(file.name)
```

This can't be done easily with the current models because the models themselves don't any session context. The same functionality would be more like:

**Comment by nitrosx** [comment]
@dylanmcreynolds , thank you for taking the time to chat today. 
With my next PR, user will be able to run the second snippet of code.
I'm wondering if there is a way to extend the Pydantic data model with additional properties and methods?
Have you come across anything like that?

**Comment by dylanmcreynolds** [comment]
I'm not positive I understand the question. Pydantic model objects extend pydantic's `BaseModel` class, you can further extend. Is that what you mean?

**Comment by nitrosx** [comment]
@dylanmcreynolds Yes. That's what I meant!!! I did not know if there are any constraint in doing that or not.
I'm not super familiar with Pydantic
